bug: llm-text-transform - error when loading file and running step

When the user loads an existing transform from Json file they are unable to run a step

browser console error:
llm.js:98  Error calling OpenAI API: TypeError: Cannot read properties of undefined (reading '0')
    at sendMessages (llm.js:93:32)
    at async runStep (transformer.js:491:30)
overrideMethod @ hook.js:608
sendMessages @ llm.js:98
await in sendMessages
runStep @ transformer.js:491
await in runStep
(anonymous) @ transformer.js:724
transformer.js:516  Error running step: TypeError: Cannot read properties of undefined (reading '0')
    at sendMessages (llm.js:93:32)
    at async runStep (transformer.js:491:30)
overrideMethod @ hook.js:608
runStep @ transformer.js:516
await in runStep
(anonymous) @ transformer.js:724


NOTE: running the step a 2nd time works so it's only an issue the first run, maybe an initial load error?

--- a/js/llm.js
+++ b/js/llm.js
@@ -100,4 +100,79 @@ async function sendMessages(messages, model = 'deepseek/deepseek-r1:free', maxTo
     }
 }
 
-export { callOpenAI, sendMessages };
+/**
+ * Fetches available models from the specified API
+ * @param {Object} settings - Required settings object containing openaiApiKey and openaiBaseUrl
+ * @returns {Promise<Array>} - Array of available model objects
+ */
+async function getAvailableModels(settings) {
+    try {
+        // Validate that settings were provided
+        if (!settings) {
+            throw new Error('Settings object is required');
+        }
+        
+        const apiKey = settings.openaiApiKey;
+        const baseUrl = settings.openaiBaseUrl || 'https://api.openai.com/v1';
+
+        if (!apiKey) {
+            throw new Error('API key is missing. Please configure it in the settings.');
+        }
+
+        // Prepare headers
+        const headers = {
+            'Content-Type': 'application/json',
+            'Authorization': `Bearer ${apiKey.trim()}`
+        };
+        
+        // Add OpenRouter specific headers if using OpenRouter
+        if (baseUrl.includes('openrouter')) {
+            headers['HTTP-Referer'] = 'https://joelmnz.github.io/ai-slop/';
+            headers['X-Title'] = 'AI Slop';
+        }
+
+        // Different endpoints depending on the API
+        let endpoint = '/models';
+        
+        // Fetch models from the API
+        const response = await fetch(`${baseUrl}${endpoint}`, {
+            method: 'GET',
+            headers: headers
+        });
+
+        // Handle response
+        if (!response.ok) {
+            const errorData = await response.json().catch(() => ({}));
+            const errorMessage = errorData.error?.message || response.statusText;
+            console.error('API Response Error:', response.status, errorMessage, errorData);
+            throw new Error(`API Error: ${errorMessage}`);
+        }
+
+        const data = await response.json();
+        
+        // Format is different between OpenAI and other providers
+        let models = [];
+        
+        if (baseUrl.includes('openrouter')) {
+            // OpenRouter format
+            models = data.data || [];
+        } else if (baseUrl.includes('openai')) {
+            // OpenAI format
+            models = data.data || [];
+        } else {
+            // Generic format - try to extract model IDs
+            models = data.data || data.models || [];
+        }
+        
+        // Return sorted list of model IDs
+        return models
+            .filter(model => !model.id.includes('deprecated') && !model.id.includes('-instruct'))
+            .map(model => model.id)
+            .sort();
+    } catch (error) {
+        console.error('Error fetching available models:', error);
+        throw error;
+    }
+}
+
+export { callOpenAI, sendMessages, getAvailableModels };

--- a/js/llm.js
+++ b/js/llm.js
@@ -160,9 +160,9 @@ async function getAvailableModels(settings) {
                 let pricingInfo = null;
                 if (model.pricing) {
                     const promptPrice = model.pricing.prompt ? 
-                        parseFloat(model.pricing.prompt) * 1000000 : null;
+                        Math.max(0, parseFloat(model.pricing.prompt) * 1000000) : null;
                     const completionPrice = model.pricing.completion ? 
-                        parseFloat(model.pricing.completion) * 1000000 : null;
+                        Math.max(0, parseFloat(model.pricing.completion) * 1000000) : null;
                     
                     if (promptPrice !== null && completionPrice !== null) {
                         pricingInfo = `$${promptPrice.toFixed(2)} in / $${completionPrice.toFixed(2)} out`;
@@ -176,9 +176,15 @@ async function getAvailableModels(settings) {
                 // Format context length if available
                 let contextLength = null;
                 if (model.context_length) {
-                    contextLength = model.context_length.toLocaleString();
+                    // Convert to "k" format (e.g., 8000 to "8k")
+                    contextLength = model.context_length >= 1000 
+                        ? `${Math.round(model.context_length / 1000)}k` 
+                        : model.context_length.toString();
                 } else if (model.top_provider?.context_length) {
-                    contextLength = model.top_provider.context_length.toLocaleString();
+                    // Convert to "k" format (e.g., 8000 to "8k")
+                    contextLength = model.top_provider.context_length >= 1000 
+                        ? `${Math.round(model.top_provider.context_length / 1000)}k` 
+                        : model.top_provider.context_length.toString();
                 }
                 
                 return {

--- a/js/llm.js
+++ b/js/llm.js
@@ -173,27 +173,57 @@ async function getAvailableModels(settings) {
                     }
                 }
                 
+                // Format context length if available
+                let contextLength = null;
+                if (model.context_length) {
+                    contextLength = model.context_length.toLocaleString();
+                } else if (model.top_provider?.context_length) {
+                    contextLength = model.top_provider.context_length.toLocaleString();
+                }
+                
                 return {
                     id: model.id || model.name || `unknown-${Math.random().toString(36).substring(2, 9)}`,
                     name: model.name || model.id || 'Unknown Model',
-                    pricing: pricingInfo
+                    pricing: pricingInfo,
+                    contextLength: contextLength,
+                    // Combine pricing and context length if both are available
+                    displayInfo: [
+                        pricingInfo,
+                        contextLength ? `${contextLength} tokens` : null
+                    ].filter(Boolean).join(' • ')
                 };
             });
         } else if (baseUrl.includes('openai')) {
             // OpenAI format only has id
-            models = (data.data || []).map(model => ({
-                id: model.id || `unknown-${Math.random().toString(36).substring(2, 9)}`,
-                name: model.id || 'Unknown Model',
-                pricing: null // OpenAI API doesn't provide pricing in the models response
-            }));
+            models = (data.data || []).map(model => {
+                // Extract context length if available
+                const contextLength = model.context_length || model.max_tokens;
+                
+                return {
+                    id: model.id || `unknown-${Math.random().toString(36).substring(2, 9)}`,
+                    name: model.id || 'Unknown Model',
+                    pricing: null, // OpenAI API doesn't provide pricing in the models response
+                    contextLength: contextLength ? contextLength.toLocaleString() : null,
+                    displayInfo: contextLength ? `${contextLength.toLocaleString()} tokens` : null
+                };
+            });
         } else {
             // Generic format - try to extract model IDs and use as names
             const rawModels = data.data || data.models || [];
-            models = rawModels.map(model => ({
-                id: model.id || model.name || `unknown-${Math.random().toString(36).substring(2, 9)}`,
-                name: model.name || model.id || 'Unknown Model',
-                pricing: null // Default to no pricing for other APIs
-            }));
+            models = rawModels.map(model => {
+                // Extract context length if available
+                const contextLength = model.context_length || 
+                                     model.max_tokens || 
+                                     model.top_provider?.context_length;
+                
+                return {
+                    id: model.id || model.name || `unknown-${Math.random().toString(36).substring(2, 9)}`,
+                    name: model.name || model.id || 'Unknown Model',
+                    pricing: null, // Default to no pricing for other APIs
+                    contextLength: contextLength ? contextLength.toLocaleString() : null,
+                    displayInfo: contextLength ? `${contextLength.toLocaleString()} tokens` : null
+                };
+            });
         }
         
         // Return sorted list of model objects with id and name, filtering out any without IDs

--- a/js/llm.js
+++ b/js/llm.js
@@ -103,7 +103,7 @@ async function sendMessages(messages, model = 'deepseek/deepseek-r1:free', maxTo
 /**
  * Fetches available models from the specified API
  * @param {Object} settings - Required settings object containing openaiApiKey and openaiBaseUrl
- * @returns {Promise<Array>} - Array of available model objects
+ * @returns {Promise<Array>} - Array of available model objects with id and name properties
  */
 async function getAvailableModels(settings) {
     try {
@@ -154,21 +154,53 @@ async function getAvailableModels(settings) {
         let models = [];
         
         if (baseUrl.includes('openrouter')) {
-            // OpenRouter format
-            models = data.data || [];
+            // OpenRouter format has name field and pricing information
+            models = (data.data || []).map(model => {
+                // Format pricing if available
+                let pricingInfo = null;
+                if (model.pricing) {
+                    const promptPrice = model.pricing.prompt ? 
+                        parseFloat(model.pricing.prompt) * 1000000 : null;
+                    const completionPrice = model.pricing.completion ? 
+                        parseFloat(model.pricing.completion) * 1000000 : null;
+                    
+                    if (promptPrice !== null && completionPrice !== null) {
+                        pricingInfo = `$${promptPrice.toFixed(2)} in / $${completionPrice.toFixed(2)} out`;
+                    } else if (promptPrice !== null) {
+                        pricingInfo = `$${promptPrice.toFixed(2)} in`;
+                    } else if (completionPrice !== null) {
+                        pricingInfo = `$${completionPrice.toFixed(2)} out`;
+                    }
+                }
+                
+                return {
+                    id: model.id || model.name || `unknown-${Math.random().toString(36).substring(2, 9)}`,
+                    name: model.name || model.id || 'Unknown Model',
+                    pricing: pricingInfo
+                };
+            });
         } else if (baseUrl.includes('openai')) {
-            // OpenAI format
-            models = data.data || [];
+            // OpenAI format only has id
+            models = (data.data || []).map(model => ({
+                id: model.id || `unknown-${Math.random().toString(36).substring(2, 9)}`,
+                name: model.id || 'Unknown Model',
+                pricing: null // OpenAI API doesn't provide pricing in the models response
+            }));
         } else {
-            // Generic format - try to extract model IDs
-            models = data.data || data.models || [];
+            // Generic format - try to extract model IDs and use as names
+            const rawModels = data.data || data.models || [];
+            models = rawModels.map(model => ({
+                id: model.id || model.name || `unknown-${Math.random().toString(36).substring(2, 9)}`,
+                name: model.name || model.id || 'Unknown Model',
+                pricing: null // Default to no pricing for other APIs
+            }));
         }
         
-        // Return sorted list of model IDs
+        // Return sorted list of model objects with id and name, filtering out any without IDs
         return models
+            .filter(model => model.id) // Ensure there's a valid ID
             .filter(model => !model.id.includes('deprecated') && !model.id.includes('-instruct'))
-            .map(model => model.id)
-            .sort();
+            .sort((a, b) => (a.name || '').localeCompare(b.name || ''));
     } catch (error) {
         console.error('Error fetching available models:', error);
         throw error;

--- a/llm-text-transform/index.html
+++ b/llm-text-transform/index.html
@@ -38,6 +38,9 @@
             <div class="title-row">
                 <label for="transformTitle">Transform Title:</label>
                 <input type="text" id="transformTitle" placeholder="Enter a title for this transformation...">
+                <a href="https://openrouter.ai/models" target="_blank" class="models-link" title="Browse OpenRouter models">
+                    <i class="fas fa-external-link-alt"></i> Models
+                </a>
             </div>
         </div>
 

--- a/llm-text-transform/index.html
+++ b/llm-text-transform/index.html
@@ -30,7 +30,6 @@
                 <button id="transformBtn"><i class="fas fa-magic"></i> Run All Steps</button>
                 <button id="cancelTransformBtn" style="display: none;"><i class="fas fa-stop"></i> Cancel</button>
                 <span id="transformProgress" style="display: none;"><i class="fas fa-spinner fa-spin"></i> Running step <span id="currentStepNum">0</span> of <span id="totalStepNum">0</span></span>
-                <button id="toggleMarkdownBtn" title="Toggles Markdown between text input and display "><i class="fas fa-code"></i> Toggle Markdown</button>
                 <button id="settings-button" title="Settings"><i class="fas fa-cog"></i></button>
             </div>
         </div>
@@ -64,7 +63,10 @@
             </div>
             
             <div class="column" id="outputColumn">
-                <label for="outputText" class="model-name-label">Output Text:</label>
+                <div class="output-header">
+                    <label for="outputText" class="model-name-label">Output Text:</label>
+                    <button id="toggleMarkdownBtn" title="Toggles Markdown between text input and display"><i class="fas fa-code"></i> Toggle Markdown</button>
+                </div>
                 <textarea id="outputText" class="llm-response-textarea" placeholder="Transformed text will appear here..." readonly></textarea>
                 
                 <div class="preview-container" style="display: none;">

--- a/llm-text-transform/styles.css
+++ b/llm-text-transform/styles.css
@@ -1022,6 +1022,10 @@ button i {
     border-bottom: 1px solid #444;
     position: relative;
     flex-shrink: 0;
+    display: flex;
+    justify-content: space-between;
+    align-items: center;
+    gap: 15px;
 }
 
 .model-search-input {
@@ -1095,16 +1099,13 @@ button i {
 }
 
 .model-selector-close {
-    position: absolute;
-    top: 15px;
-    right: 15px;
+    position: relative; /* Changed from absolute */
     background: #333;
     border: none;
     color: #fff;
     cursor: pointer;
     padding: 8px;
     font-size: 1.2rem;
-    z-index: 1;
     border-radius: 50%;
     width: 36px;
     height: 36px;
@@ -1112,6 +1113,7 @@ button i {
     align-items: center;
     justify-content: center;
     transition: background-color 0.2s;
+    flex-shrink: 0; /* Prevent shrinking */
 }
 
 .model-selector-close:hover {

--- a/llm-text-transform/styles.css
+++ b/llm-text-transform/styles.css
@@ -770,6 +770,19 @@ button i {
     background-color: var(--primary-color-hover);
 }
 
+/* Output header styling */
+.output-header {
+    display: flex;
+    justify-content: space-between;
+    align-items: center;
+    margin-bottom: 5px;
+}
+
+.output-header button {
+    padding: 5px 10px;
+    font-size: 0.8rem;
+}
+
 /* Print styles - updated for proper A4 printing */
 @media print {
     @page {

--- a/llm-text-transform/styles.css
+++ b/llm-text-transform/styles.css
@@ -1105,6 +1105,16 @@ button i {
 .model-name {
     font-weight: 500;
     font-size: 1.1rem;
+    margin-right: auto;
+}
+
+.model-context {
+    font-size: 0.9rem;
+    color: #aaa;
+    margin-left: 15px;
+    padding-left: 15px;
+    border-left: 1px solid #555;
+    white-space: nowrap;
 }
 
 .model-pricing {

--- a/llm-text-transform/styles.css
+++ b/llm-text-transform/styles.css
@@ -551,12 +551,20 @@ button i {
     margin-bottom: 10px;
 }
 
-/* Adjust step model container for autocomplete */
+/* Adjust step model container for autocomplete - update */
 .step-model-container {
     position: relative;
     display: flex;
-    justify-content: space-between;
     align-items: center;
+    justify-content: space-between;
+}
+
+/* Step buttons container styling */
+.step-buttons-container {
+    display: flex;
+    justify-content: flex-end;
+    gap: 10px;
+    margin-top: 10px;
 }
 
 .step-model-label {

--- a/llm-text-transform/styles.css
+++ b/llm-text-transform/styles.css
@@ -1028,9 +1028,17 @@ button i {
     gap: 15px;
 }
 
+.search-input-wrapper {
+    position: relative;
+    flex-grow: 1;
+    display: flex;
+    align-items: center;
+}
+
 .model-search-input {
     width: 100%;
     padding: 12px 15px;
+    padding-right: 40px; /* Space for the clear button */
     border: 1px solid #444;
     border-radius: 4px;
     background-color: #2a2a2a;
@@ -1038,10 +1046,27 @@ button i {
     font-size: 1rem;
 }
 
-.model-search-input:focus {
-    outline: none;
-    border-color: #6610f2;
-    box-shadow: 0 0 0 2px rgba(102, 16, 242, 0.25);
+.search-clear-btn {
+    position: absolute;
+    right: 10px;
+    background: transparent;
+    border: none;
+    color: #aaa;
+    cursor: pointer;
+    padding: 5px;
+    font-size: 0.9rem;
+    display: flex;
+    align-items: center;
+    justify-content: center;
+    border-radius: 50%;
+    width: 24px;
+    height: 24px;
+    transition: all 0.2s;
+}
+
+.search-clear-btn:hover {
+    color: #fff;
+    background-color: rgba(255, 255, 255, 0.1);
 }
 
 .models-list-container {

--- a/llm-text-transform/styles.css
+++ b/llm-text-transform/styles.css
@@ -72,6 +72,34 @@ h1 {
     margin-bottom: 0;
 }
 
+/* Models link styling */
+.models-link {
+    color: #64d8ff;
+    text-decoration: none;
+    display: flex;
+    align-items: center;
+    gap: 5px;
+    padding: 5px 10px;
+    border-radius: 4px;
+    transition: all 0.2s ease;
+    font-size: 0.9rem;
+    white-space: nowrap;
+}
+
+.models-link:hover {
+    background-color: rgba(100, 216, 255, 0.1);
+    color: #8ae7ff;
+}
+
+.models-link:visited {
+    color: #b388ff;
+}
+
+.models-link:visited:hover {
+    background-color: rgba(179, 136, 255, 0.1);
+    color: #cfb6ff;
+}
+
 label { /* General label styling */
     font-weight: bold;
     margin-bottom: 5px;

--- a/llm-text-transform/styles.css
+++ b/llm-text-transform/styles.css
@@ -713,7 +713,7 @@ button i {
     justify-content: space-between;
     align-items: center;
     padding: 15px 20px;
-    background-color: #3700b3;
+    background-color: var(--dark-highlight, #3498db);
     border-bottom: 1px solid var(--dark-border);
 }
 
@@ -743,6 +743,7 @@ button i {
     padding: 20px;
     overflow-y: auto;
     max-height: calc(80vh - 60px);
+    background-color: #222; /* Changed from var(--dark-accent) to a specific dark color */
 }
 
 .response-content {

--- a/llm-text-transform/styles.css
+++ b/llm-text-transform/styles.css
@@ -551,7 +551,9 @@ button i {
     margin-bottom: 10px;
 }
 
+/* Adjust step model container for autocomplete */
 .step-model-container {
+    position: relative;
     display: flex;
     justify-content: space-between;
     align-items: center;
@@ -578,7 +580,7 @@ button i {
     background-color: #2a2a2a;
     color: #e0e0e0;
     border: 1px solid #444;
-    border-radius: 4px;
+    border-radius: 4px 0 0 4px;
     padding: 8px;
     font-size: 0.9rem;
 }
@@ -946,4 +948,117 @@ button i {
     margin: 24px 0;
     background-color: #e1e4e8;
     border: 0;
+}
+
+/* Remove autocomplete styles - We're no longer using autocomplete */
+.autocomplete-container, .autocomplete-list, .autocomplete-list li {
+    display: none;
+}
+
+/* Model input group styling */
+.model-input-group {
+    display: flex;
+    flex-grow: 1;
+}
+
+/* Model lookup button styling */
+.model-lookup-btn {
+    background-color: #6610f2;
+    color: white;
+    border: none;
+    border-radius: 0 4px 4px 0;
+    padding: 0 10px;
+    cursor: pointer;
+    font-size: 0.9rem;
+    transition: background-color 0.2s;
+}
+
+.model-lookup-btn:hover {
+    background-color: #5a00e0;
+}
+
+/* Model selector popup styling */
+.model-selector-popup {
+    position: absolute;
+    z-index: 1000;
+    background-color: #222;
+    border: 1px solid #444;
+    border-radius: 4px;
+    box-shadow: 0 5px 15px rgba(0, 0, 0, 0.5);
+    display: none;
+    padding-bottom: 5px;
+}
+
+.model-search-container {
+    padding: 10px;
+    border-bottom: 1px solid #444;
+    position: relative;
+}
+
+.model-search-input {
+    width: 100%;
+    padding: 8px 10px;
+    border: 1px solid #444;
+    border-radius: 4px;
+    background-color: #2a2a2a;
+    color: #e0e0e0;
+    font-size: 0.9rem;
+}
+
+.model-search-input:focus {
+    outline: none;
+    border-color: #6610f2;
+    box-shadow: 0 0 0 2px rgba(102, 16, 242, 0.25);
+}
+
+.models-list-container {
+    max-height: 300px;
+    overflow-y: auto;
+    padding: 5px 0;
+}
+
+.models-list {
+    list-style: none;
+    padding: 0;
+    margin: 0;
+}
+
+.model-item {
+    padding: 8px 15px;
+    cursor: pointer;
+    transition: background-color 0.2s;
+    color: #e0e0e0;
+}
+
+.model-item:hover {
+    background-color: #3a3a3a;
+}
+
+.model-item.selected {
+    background-color: rgba(102, 16, 242, 0.2);
+    border-left: 3px solid #6610f2;
+}
+
+.no-models-message {
+    padding: 15px;
+    text-align: center;
+    color: #999;
+    font-style: italic;
+}
+
+.model-selector-close {
+    position: absolute;
+    top: 10px;
+    right: 10px;
+    background: transparent;
+    border: none;
+    color: #aaa;
+    cursor: pointer;
+    padding: 5px;
+    font-size: 1rem;
+    z-index: 1;
+}
+
+.model-selector-close:hover {
+    color: #fff;
 }

--- a/llm-text-transform/styles.css
+++ b/llm-text-transform/styles.css
@@ -987,30 +987,51 @@ button i {
 
 /* Model selector popup styling */
 .model-selector-popup {
-    position: absolute;
+    position: fixed;
     z-index: 1000;
     background-color: #222;
     border: 1px solid #444;
-    border-radius: 4px;
-    box-shadow: 0 5px 15px rgba(0, 0, 0, 0.5);
+    border-radius: 8px;
+    box-shadow: 0 10px 25px rgba(0, 0, 0, 0.5);
     display: none;
-    padding-bottom: 5px;
+    width: 90%;
+    max-width: 800px;
+    height: 90vh;
+    top: 50%;
+    left: 50%;
+    transform: translate(-50%, -50%);
+    padding: 0;
+    overflow: hidden;
+    display: flex;
+    flex-direction: column;
+}
+
+.model-selector-popup::before {
+    content: '';
+    position: fixed;
+    top: 0;
+    left: 0;
+    right: 0;
+    bottom: 0;
+    background-color: rgba(0, 0, 0, 0.7);
+    z-index: -1;
 }
 
 .model-search-container {
-    padding: 10px;
+    padding: 20px;
     border-bottom: 1px solid #444;
     position: relative;
+    flex-shrink: 0;
 }
 
 .model-search-input {
     width: 100%;
-    padding: 8px 10px;
+    padding: 12px 15px;
     border: 1px solid #444;
     border-radius: 4px;
     background-color: #2a2a2a;
     color: #e0e0e0;
-    font-size: 0.9rem;
+    font-size: 1rem;
 }
 
 .model-search-input:focus {
@@ -1020,9 +1041,9 @@ button i {
 }
 
 .models-list-container {
-    max-height: 300px;
+    flex: 1;
     overflow-y: auto;
-    padding: 5px 0;
+    padding: 0;
 }
 
 .models-list {
@@ -1035,10 +1056,11 @@ button i {
     display: flex;
     justify-content: space-between;
     align-items: center;
-    padding: 8px 12px;
+    padding: 12px 20px;
     cursor: pointer;
     transition: background-color 0.2s;
     color: #e0e0e0;
+    border-bottom: 1px solid #333;
 }
 
 .model-item:hover {
@@ -1047,42 +1069,58 @@ button i {
 
 .model-item.selected {
     background-color: rgba(102, 16, 242, 0.2);
-    border-left: 3px solid #6610f2;
+    border-left: 4px solid #6610f2;
 }
 
 .model-name {
     font-weight: 500;
+    font-size: 1.1rem;
 }
 
 .model-pricing {
-    font-size: 0.85em;
+    font-size: 0.9rem;
     color: #aaa;
-    margin-left: 10px;
-    padding-left: 8px;
+    margin-left: 15px;
+    padding-left: 15px;
     border-left: 1px solid #555;
     white-space: nowrap;
 }
 
 .no-models-message {
-    padding: 15px;
+    padding: 30px;
     text-align: center;
     color: #999;
     font-style: italic;
+    font-size: 1.1rem;
 }
 
 .model-selector-close {
     position: absolute;
-    top: 10px;
-    right: 10px;
-    background: transparent;
+    top: 15px;
+    right: 15px;
+    background: #333;
     border: none;
-    color: #aaa;
+    color: #fff;
     cursor: pointer;
-    padding: 5px;
-    font-size: 1rem;
+    padding: 8px;
+    font-size: 1.2rem;
     z-index: 1;
+    border-radius: 50%;
+    width: 36px;
+    height: 36px;
+    display: flex;
+    align-items: center;
+    justify-content: center;
+    transition: background-color 0.2s;
 }
 
 .model-selector-close:hover {
-    color: #fff;
+    background-color: #555;
+}
+
+.loading-models {
+    padding: 30px;
+    text-align: center;
+    color: #e0e0e0;
+    font-size: 1.1rem;
 }

--- a/llm-text-transform/styles.css
+++ b/llm-text-transform/styles.css
@@ -1024,7 +1024,10 @@ button i {
 }
 
 .model-item {
-    padding: 8px 15px;
+    display: flex;
+    justify-content: space-between;
+    align-items: center;
+    padding: 8px 12px;
     cursor: pointer;
     transition: background-color 0.2s;
     color: #e0e0e0;
@@ -1037,6 +1040,19 @@ button i {
 .model-item.selected {
     background-color: rgba(102, 16, 242, 0.2);
     border-left: 3px solid #6610f2;
+}
+
+.model-name {
+    font-weight: 500;
+}
+
+.model-pricing {
+    font-size: 0.85em;
+    color: #aaa;
+    margin-left: 10px;
+    padding-left: 8px;
+    border-left: 1px solid #555;
+    white-space: nowrap;
 }
 
 .no-models-message {

--- a/llm-text-transform/transformer.js
+++ b/llm-text-transform/transformer.js
@@ -1,13 +1,18 @@
 document.addEventListener('DOMContentLoaded', async () => {
     // Try to dynamically import the callOpenAI function
     let sendMessages;
+    let getAvailableModels;
     try {
         const module = await import('../js/llm.js');
         sendMessages = module.sendMessages;
+        getAvailableModels = module.getAvailableModels;
     } catch (error) {
         console.error('Error importing llm.js:', error);
         // Create fallback if import fails
         sendMessages = async (messages, model, maxTokens, settings) => {
+            throw new Error('LLM API module could not be loaded. Please check the console for details.');
+        };
+        getAvailableModels = async (settings) => {
             throw new Error('LLM API module could not be loaded. Please check the console for details.');
         };
     }
@@ -41,6 +46,7 @@ document.addEventListener('DOMContentLoaded', async () => {
     let stepCounter = 0;
     let isTransforming = false;
     let cancelTransformation = false;
+    let availableModels = []; // New state for storing available models
 
     // --- API Settings Cache ---
     let apiSettings = null;
@@ -62,6 +68,198 @@ document.addEventListener('DOMContentLoaded', async () => {
         }
         
         return apiSettings;
+    }
+
+    // --- Fetch available models ---
+    async function fetchAvailableModels() {
+        try {
+            if (typeof window.aiSettings === 'undefined') {
+                throw new Error('Settings manager is not available');
+            }
+            
+            const settings = await getApiSettings();
+            availableModels = await getAvailableModels(settings);
+            
+            return availableModels;
+        } catch (error) {
+            console.error('Error fetching available models:', error);
+            setStatus(`Failed to load model list: ${error.message}`, 'error', 5000);
+            return [];
+        }
+    }
+    
+    // Model selector popup functions
+    let modelSelectorPopup = null;
+    let currentModelInput = null;
+    
+    // Create model selector popup if it doesn't exist
+    function createModelSelectorPopup() {
+        if (modelSelectorPopup) return modelSelectorPopup;
+        
+        // Create popup container
+        modelSelectorPopup = document.createElement('div');
+        modelSelectorPopup.className = 'model-selector-popup';
+        modelSelectorPopup.style.display = 'none';
+        
+        // Add search input
+        const searchContainer = document.createElement('div');
+        searchContainer.className = 'model-search-container';
+        
+        const searchInput = document.createElement('input');
+        searchInput.type = 'text';
+        searchInput.className = 'model-search-input';
+        searchInput.placeholder = 'Search models...';
+        
+        searchContainer.appendChild(searchInput);
+        modelSelectorPopup.appendChild(searchContainer);
+        
+        // Add models list container
+        const modelsListContainer = document.createElement('div');
+        modelsListContainer.className = 'models-list-container';
+        modelSelectorPopup.appendChild(modelsListContainer);
+        
+        // Add close button
+        const closeButton = document.createElement('button');
+        closeButton.className = 'model-selector-close';
+        closeButton.innerHTML = '<i class="fas fa-times"></i>';
+        modelSelectorPopup.appendChild(closeButton);
+        
+        // Add event listeners
+        searchInput.addEventListener('input', function() {
+            const searchTerm = this.value.toLowerCase();
+            updateModelsList(searchTerm);
+        });
+        
+        closeButton.addEventListener('click', function() {
+            hideModelSelector();
+        });
+        
+        // Close when clicking outside
+        document.addEventListener('click', function(e) {
+            if (modelSelectorPopup && modelSelectorPopup.style.display !== 'none') {
+                // Check if click is outside the popup
+                if (!modelSelectorPopup.contains(e.target) && 
+                    !e.target.classList.contains('model-lookup-btn') &&
+                    !e.target.closest('.model-lookup-btn')) {
+                    hideModelSelector();
+                }
+            }
+        });
+        
+        // Add to body
+        document.body.appendChild(modelSelectorPopup);
+        return modelSelectorPopup;
+    }
+    
+    // Show model selector popup
+    function showModelSelector(modelInput) {
+        // Store reference to the input field
+        currentModelInput = modelInput;
+        
+        // Create popup if it doesn't exist
+        const popup = createModelSelectorPopup();
+        
+        // Position popup near the input
+        const inputRect = modelInput.getBoundingClientRect();
+        popup.style.top = `${inputRect.bottom + window.scrollY + 5}px`;
+        popup.style.left = `${inputRect.left + window.scrollX}px`;
+        popup.style.width = `${Math.max(300, inputRect.width)}px`;
+        
+        // Update models list with any existing filter value
+        const searchInput = popup.querySelector('.model-search-input');
+        searchInput.value = modelInput.value; // Set search to current input value
+        updateModelsList(searchInput.value.toLowerCase());
+        
+        // Show popup
+        popup.style.display = 'block';
+        
+        // Focus search input
+        setTimeout(() => {
+            searchInput.focus();
+        }, 100);
+    }
+    
+    // Hide model selector popup
+    function hideModelSelector() {
+        if (modelSelectorPopup) {
+            modelSelectorPopup.style.display = 'none';
+            currentModelInput = null;
+        }
+    }
+    
+    // Update models list based on search term
+    function updateModelsList(searchTerm = '') {
+        if (!modelSelectorPopup) return;
+        
+        const modelsListContainer = modelSelectorPopup.querySelector('.models-list-container');
+        modelsListContainer.innerHTML = '';
+        
+        // Filter models by search term
+        const filteredModels = availableModels.filter(model => 
+            model.toLowerCase().includes(searchTerm)
+        );
+        
+        if (filteredModels.length === 0) {
+            const noResults = document.createElement('div');
+            noResults.className = 'no-models-message';
+            noResults.textContent = searchTerm ? 
+                'No models matching your search' : 
+                'No models available. Check your API settings.';
+            modelsListContainer.appendChild(noResults);
+            return;
+        }
+        
+        // Create list of models
+        const modelsList = document.createElement('ul');
+        modelsList.className = 'models-list';
+        
+        filteredModels.forEach(model => {
+            const modelItem = document.createElement('li');
+            modelItem.className = 'model-item';
+            modelItem.textContent = model;
+            
+            // Highlight current selection
+            if (currentModelInput && currentModelInput.value === model) {
+                modelItem.classList.add('selected');
+            }
+            
+            // Add click handler
+            modelItem.addEventListener('click', function() {
+                if (currentModelInput) {
+                    currentModelInput.value = model;
+                    // Trigger input event to update run button state
+                    currentModelInput.dispatchEvent(new Event('input'));
+                }
+                hideModelSelector();
+            });
+            
+            modelsList.appendChild(modelItem);
+        });
+        
+        modelsListContainer.appendChild(modelsList);
+    }
+    
+    // Function to show the LLM response popup
+    function showResponsePopup(responseText) {
+        // Create popup if it doesn't exist
+        let overlay = document.getElementById('responsePopupOverlay');
+        if (!overlay) {
+            const popupTemplate = templates.responsePopup();
+            overlay = renderTemplate(popupTemplate, document.body);
+            
+            // Add close button event listener
+            const closeBtn = overlay.querySelector('.close-popup-btn');
+            closeBtn.addEventListener('click', () => {
+                overlay.style.display = 'none';
+            });
+        }
+        
+        // Set response content
+        const responseContent = overlay.querySelector('#responseContent');
+        responseContent.textContent = responseText;
+        
+        // Show the popup
+        overlay.style.display = 'flex';
     }
 
     // --- Configure marked options ---
@@ -87,11 +285,6 @@ document.addEventListener('DOMContentLoaded', async () => {
             classes: ['transform-step'],
             attributes: { id: `step${id}`, 'data-step-id': id },
             children: [
-                {
-                    type: 'div',
-                    classes: ['step-number'],
-                    text: id
-                },
                 {
                     type: 'div',
                     classes: ['llm-response-icon'],
@@ -137,14 +330,29 @@ document.addEventListener('DOMContentLoaded', async () => {
                             text: 'Model:'
                         },
                         {
-                            type: 'input',
-                            classes: ['step-model-input'],
-                            attributes: { 
-                                id: `stepModel${id}`,
-                                type: 'text',
-                                placeholder: 'e.g., gpt-4o, claude-3, llama, etc.',
-                                value: data?.model || ''
-                            }
+                            type: 'div',
+                            classes: ['model-input-group'],
+                            children: [
+                                {
+                                    type: 'input',
+                                    classes: ['step-model-input'],
+                                    attributes: { 
+                                        id: `stepModel${id}`,
+                                        type: 'text',
+                                        placeholder: 'e.g., gpt-4o, claude-3, llama, etc.',
+                                        value: data?.model || ''
+                                    }
+                                },
+                                {
+                                    type: 'button',
+                                    classes: ['model-lookup-btn'],
+                                    attributes: { 
+                                        'data-input-id': `stepModel${id}`,
+                                        'title': 'Look up available models'
+                                    },
+                                    html: '<i class="fas fa-search"></i>'
+                                }
+                            ]
                         },
                         {
                             type: 'button',
@@ -156,7 +364,7 @@ document.addEventListener('DOMContentLoaded', async () => {
                             type: 'button',
                             classes: ['run-step-btn'],
                             attributes: { 'data-step-id': id },
-                            html: '<i class="fas fa-play"></i> Run Step'
+                            html: '<i class="fas fa-play"></i> Run'
                         }
                     ]
                 }
@@ -306,7 +514,6 @@ document.addEventListener('DOMContentLoaded', async () => {
         const steps = stepsContainer.querySelectorAll('.transform-step');
         steps.forEach((step, index) => {
             const stepNumber = index + 1;
-            step.querySelector('.step-number').textContent = stepNumber;
             step.querySelector('.step-instructions-label').textContent = `Step ${stepNumber} Instructions:`;
         });
     }
@@ -520,7 +727,7 @@ document.addEventListener('DOMContentLoaded', async () => {
             const step = document.querySelector(`.transform-step[data-step-id="${stepId}"]`);
             if (step) {
                 const runButton = step.querySelector('.run-step-btn');
-                runButton.innerHTML = '<i class="fas fa-play"></i> Run Step';
+                runButton.innerHTML = '<i class="fas fa-play"></i> Run';
                 runButton.disabled = false;
             }
         }
@@ -710,6 +917,24 @@ document.addEventListener('DOMContentLoaded', async () => {
                     setStatus('No LLM response available for this step', 'info');
                 }
             }
+        }
+        
+        // Handle model lookup button clicks
+        if (event.target.classList.contains('model-lookup-btn') || 
+            event.target.closest('.model-lookup-btn')) {
+            
+            const button = event.target.classList.contains('model-lookup-btn') ? 
+                event.target : event.target.closest('.model-lookup-btn');
+            
+            const inputId = button.getAttribute('data-input-id');
+            if (inputId) {
+                const input = document.getElementById(inputId);
+                if (input) {
+                    showModelSelector(input);
+                }
+            }
+            // Prevent event from bubbling up to document click handler
+            event.stopPropagation();
         }
         
         // Handle run step button clicks
@@ -927,4 +1152,9 @@ document.addEventListener('DOMContentLoaded', async () => {
 
     // Initialize the page
     resetToDefaultState();
+    
+    // Fetch available models after page initialization
+    fetchAvailableModels().catch(error => {
+        console.error('Failed to initialize models:', error);
+    });
 });

--- a/llm-text-transform/transformer.js
+++ b/llm-text-transform/transformer.js
@@ -391,7 +391,13 @@ document.addEventListener('DOMContentLoaded', async () => {
                                     html: '<i class="fas fa-search"></i>'
                                 }
                             ]
-                        },
+                        }
+                    ]
+                },
+                {
+                    type: 'div',
+                    classes: ['step-buttons-container'],
+                    children: [
                         {
                             type: 'button',
                             classes: ['remove-step-btn'],

--- a/llm-text-transform/transformer.js
+++ b/llm-text-transform/transformer.js
@@ -158,10 +158,8 @@ document.addEventListener('DOMContentLoaded', async () => {
         // Create popup if it doesn't exist
         const popup = createModelSelectorPopup();
         
-        // Position popup near the input
-        const inputRect = modelInput.getBoundingClientRect();
-        popup.style.top = `${inputRect.bottom + window.scrollY + 5}px`;
-        popup.style.left = `${inputRect.left + window.scrollX}px`;
+        // No need to position near input as it's now centered
+        // Just display the popup
         
         // Check if models are cached, fetch if needed
         if (availableModels.length === 0) {
@@ -171,7 +169,7 @@ document.addEventListener('DOMContentLoaded', async () => {
                 modelsListContainer.innerHTML = '<div class="loading-models">Loading models...</div>';
                 
                 // Show popup while loading
-                popup.style.display = 'block';
+                popup.style.display = 'flex'; // Changed from 'block' to 'flex'
                 
                 // Fetch and cache models
                 await fetchAvailableModels();
@@ -181,7 +179,7 @@ document.addEventListener('DOMContentLoaded', async () => {
             }
         } else {
             // Show popup immediately if models are already cached
-            popup.style.display = 'block';
+            popup.style.display = 'flex'; // Changed from 'block' to 'flex'
         }
         
         // Update models list with any existing filter value
@@ -193,6 +191,9 @@ document.addEventListener('DOMContentLoaded', async () => {
         setTimeout(() => {
             searchInput.focus();
         }, 100);
+        
+        // Add event listener to close popup when clicking outside
+        // This is already handled in the createModelSelectorPopup function
     }
     
     // Hide model selector popup

--- a/llm-text-transform/transformer.js
+++ b/llm-text-transform/transformer.js
@@ -152,7 +152,7 @@ document.addEventListener('DOMContentLoaded', async () => {
     }
     
     // Show model selector popup
-    function showModelSelector(modelInput) {
+    async function showModelSelector(modelInput) {
         // Store reference to the input field
         currentModelInput = modelInput;
         
@@ -165,13 +165,31 @@ document.addEventListener('DOMContentLoaded', async () => {
         popup.style.left = `${inputRect.left + window.scrollX}px`;
         popup.style.width = `${Math.max(300, inputRect.width)}px`;
         
+        // Check if models are cached, fetch if needed
+        if (availableModels.length === 0) {
+            try {
+                // Show loading indicator in popup
+                const modelsListContainer = popup.querySelector('.models-list-container');
+                modelsListContainer.innerHTML = '<div class="loading-models">Loading models...</div>';
+                
+                // Show popup while loading
+                popup.style.display = 'block';
+                
+                // Fetch and cache models
+                await fetchAvailableModels();
+            } catch (error) {
+                console.error('Error fetching models:', error);
+                setStatus(`Failed to load models: ${error.message}`, 'error', 5000);
+            }
+        } else {
+            // Show popup immediately if models are already cached
+            popup.style.display = 'block';
+        }
+        
         // Update models list with any existing filter value
         const searchInput = popup.querySelector('.model-search-input');
         searchInput.value = modelInput.value; // Set search to current input value
         updateModelsList(searchInput.value.toLowerCase());
-        
-        // Show popup
-        popup.style.display = 'block';
         
         // Focus search input
         setTimeout(() => {
@@ -1000,6 +1018,8 @@ document.addEventListener('DOMContentLoaded', async () => {
             if (typeof window.aiSettings !== 'undefined') {
                 // Reset API settings cache when opening settings
                 apiSettings = null;
+                // Also clear cached models since settings might affect available models
+                availableModels = [];
                 window.aiSettings.openSettings();
             } else {
                 setStatus('Settings manager is not available', 'error');
@@ -1152,9 +1172,4 @@ document.addEventListener('DOMContentLoaded', async () => {
 
     // Initialize the page
     resetToDefaultState();
-    
-    // Fetch available models after page initialization
-    fetchAvailableModels().catch(error => {
-        console.error('Failed to initialize models:', error);
-    });
 });

--- a/llm-text-transform/transformer.js
+++ b/llm-text-transform/transformer.js
@@ -109,19 +109,20 @@ document.addEventListener('DOMContentLoaded', async () => {
         searchInput.className = 'model-search-input';
         searchInput.placeholder = 'Search models...';
         
+        // Add close button to search container instead of directly to popup
+        const closeButton = document.createElement('button');
+        closeButton.className = 'model-selector-close';
+        closeButton.innerHTML = '<i class="fas fa-times"></i>';
+        
+        // Add elements to container
         searchContainer.appendChild(searchInput);
+        searchContainer.appendChild(closeButton);
         modelSelectorPopup.appendChild(searchContainer);
         
         // Add models list container
         const modelsListContainer = document.createElement('div');
         modelsListContainer.className = 'models-list-container';
         modelSelectorPopup.appendChild(modelsListContainer);
-        
-        // Add close button
-        const closeButton = document.createElement('button');
-        closeButton.className = 'model-selector-close';
-        closeButton.innerHTML = '<i class="fas fa-times"></i>';
-        modelSelectorPopup.appendChild(closeButton);
         
         // Add event listeners
         searchInput.addEventListener('input', function() {

--- a/llm-text-transform/transformer.js
+++ b/llm-text-transform/transformer.js
@@ -269,6 +269,14 @@ document.addEventListener('DOMContentLoaded', async () => {
             nameSpan.textContent = model.name || model.id;
             modelItem.appendChild(nameSpan);
             
+            // Add context length info if available
+            if (model.contextLength) {
+                const contextSpan = document.createElement('span');
+                contextSpan.className = 'model-context';
+                contextSpan.textContent = `${model.contextLength} context`;
+                modelItem.appendChild(contextSpan);
+            }
+            
             // Add pricing info if available
             if (model.pricing) {
                 const pricingSpan = document.createElement('span');
@@ -789,7 +797,7 @@ document.addEventListener('DOMContentLoaded', async () => {
             
             // Show the response icon
             const responseIcon = step.querySelector('.llm-response-icon');
-            responseIcon.style.display = 'block';
+            responseIcon.style.display = 'flex';
             
             // Update the output text area
             outputText.value = response;

--- a/llm-text-transform/transformer.js
+++ b/llm-text-transform/transformer.js
@@ -46,7 +46,7 @@ document.addEventListener('DOMContentLoaded', async () => {
     let stepCounter = 0;
     let isTransforming = false;
     let cancelTransformation = false;
-    let availableModels = []; // New state for storing available models
+    let availableModels = []; // Will now store objects with {id, name} instead of just strings
 
     // --- API Settings Cache ---
     let apiSettings = null;
@@ -212,10 +212,13 @@ document.addEventListener('DOMContentLoaded', async () => {
         const modelsListContainer = modelSelectorPopup.querySelector('.models-list-container');
         modelsListContainer.innerHTML = '';
         
-        // Filter models by search term
-        const filteredModels = availableModels.filter(model => 
-            model.toLowerCase().includes(searchTerm)
-        );
+        // Filter models by search term - now searching in both name and id with null checks
+        const filteredModels = availableModels.filter(model => {
+            const modelName = model.name || '';
+            const modelId = model.id || '';
+            return modelName.toLowerCase().includes(searchTerm) || 
+                   modelId.toLowerCase().includes(searchTerm);
+        });
         
         if (filteredModels.length === 0) {
             const noResults = document.createElement('div');
@@ -232,19 +235,38 @@ document.addEventListener('DOMContentLoaded', async () => {
         modelsList.className = 'models-list';
         
         filteredModels.forEach(model => {
+            // Skip invalid models
+            if (!model.id) return;
+            
             const modelItem = document.createElement('li');
             modelItem.className = 'model-item';
-            modelItem.textContent = model;
+            
+            // Create model name span
+            const nameSpan = document.createElement('span');
+            nameSpan.className = 'model-name';
+            nameSpan.textContent = model.name || model.id;
+            modelItem.appendChild(nameSpan);
+            
+            // Add pricing info if available
+            if (model.pricing) {
+                const pricingSpan = document.createElement('span');
+                pricingSpan.className = 'model-pricing';
+                pricingSpan.textContent = model.pricing;
+                modelItem.appendChild(pricingSpan);
+            }
+            
+            // Store model ID as data attribute
+            modelItem.setAttribute('data-model-id', model.id);
             
             // Highlight current selection
-            if (currentModelInput && currentModelInput.value === model) {
+            if (currentModelInput && currentModelInput.value === model.id) {
                 modelItem.classList.add('selected');
             }
             
             // Add click handler
             modelItem.addEventListener('click', function() {
                 if (currentModelInput) {
-                    currentModelInput.value = model;
+                    currentModelInput.value = model.id; // Set the ID as the value
                     // Trigger input event to update run button state
                     currentModelInput.dispatchEvent(new Event('input'));
                 }

--- a/llm-text-transform/transformer.js
+++ b/llm-text-transform/transformer.js
@@ -824,7 +824,7 @@ document.addEventListener('DOMContentLoaded', async () => {
 
             // Generate default filename from the transform title
             const safeTitle = transformData.title.replace(/[^a-z0-9]/gi, '_').toLowerCase();
-            const defaultFilename = `${safeTitle || 'llm_transform'}`;
+            const defaultFilename = `lmtx-${safeTitle || 'llm_transform'}`;
             
             // Prompt user for filename
             const userFilename = prompt("Enter filename for saving (without extension):", defaultFilename);

--- a/llm-text-transform/transformer.js
+++ b/llm-text-transform/transformer.js
@@ -302,7 +302,7 @@ document.addEventListener('DOMContentLoaded', async () => {
     }
     
     // Function to show the LLM response popup
-    function showResponsePopup(responseText) {
+    function showResponsePopup(responseText, stepId) {
         // Create popup if it doesn't exist
         let overlay = document.getElementById('responsePopupOverlay');
         if (!overlay) {
@@ -314,6 +314,24 @@ document.addEventListener('DOMContentLoaded', async () => {
             closeBtn.addEventListener('click', () => {
                 overlay.style.display = 'none';
             });
+        }
+        
+        // Get step number from step ID
+        let title = "LLM Response";
+        if (stepId) {
+            const step = document.querySelector(`.transform-step[data-step-id="${stepId}"]`);
+            if (step) {
+                const stepLabel = step.querySelector('.step-instructions-label').textContent;
+                // Extract the step number from the label (e.g., "Step 3 Instructions:")
+                const stepNumber = stepLabel.match(/Step\s+(\d+)/i)?.[1] || stepId;
+                title = `Step ${stepNumber} Response`;
+            }
+        }
+        
+        // Set the title
+        const popupTitle = overlay.querySelector('.response-popup-header h3');
+        if (popupTitle) {
+            popupTitle.textContent = title;
         }
         
         // Set response content
@@ -353,7 +371,7 @@ document.addEventListener('DOMContentLoaded', async () => {
                     attributes: {
                         'data-step-id': id,
                         'title': 'View LLM Response',
-                        'style': data?.llmResponse ? 'display: block;' : 'display: none;'
+                        'style': data?.llmResponse ? 'display: flex;' : 'display: none;'
                     },
                     html: '<i class="fas fa-robot"></i>'
                 },
@@ -980,7 +998,7 @@ document.addEventListener('DOMContentLoaded', async () => {
                 const step = document.querySelector(`.transform-step[data-step-id="${stepId}"]`);
                 const responseText = step.querySelector('.hidden-llm-response').value;
                 if (responseText) {
-                    showResponsePopup(responseText);
+                    showResponsePopup(responseText, stepId);
                 } else {
                     setStatus('No LLM response available for this step', 'info');
                 }

--- a/llm-text-transform/transformer.js
+++ b/llm-text-transform/transformer.js
@@ -38,7 +38,6 @@ document.addEventListener('DOMContentLoaded', async () => {
     const copyBtn = document.getElementById('copyBtn');
     const printBtn = document.getElementById('printBtn');
     const toastContainer = document.getElementById('toastContainer');
-    const comparisonMeta = document.querySelector('.comparison-meta');
     const settingsBtn = document.getElementById('settings-button');
 
     // --- State ---
@@ -163,7 +162,6 @@ document.addEventListener('DOMContentLoaded', async () => {
         const inputRect = modelInput.getBoundingClientRect();
         popup.style.top = `${inputRect.bottom + window.scrollY + 5}px`;
         popup.style.left = `${inputRect.left + window.scrollX}px`;
-        popup.style.width = `${Math.max(300, inputRect.width)}px`;
         
         // Check if models are cached, fetch if needed
         if (availableModels.length === 0) {

--- a/llm-text-transform/transformer.js
+++ b/llm-text-transform/transformer.js
@@ -104,18 +104,33 @@ document.addEventListener('DOMContentLoaded', async () => {
         const searchContainer = document.createElement('div');
         searchContainer.className = 'model-search-container';
         
+        // Create search input wrapper (for containing input and clear button)
+        const searchInputWrapper = document.createElement('div');
+        searchInputWrapper.className = 'search-input-wrapper';
+        
         const searchInput = document.createElement('input');
         searchInput.type = 'text';
         searchInput.className = 'model-search-input';
         searchInput.placeholder = 'Search models...';
         
-        // Add close button to search container instead of directly to popup
+        // Create clear button for search input
+        const clearButton = document.createElement('button');
+        clearButton.type = 'button';
+        clearButton.className = 'search-clear-btn';
+        clearButton.innerHTML = '<i class="fas fa-times"></i>';
+        clearButton.title = 'Clear search';
+        
+        // Add searchInput and clearButton to wrapper
+        searchInputWrapper.appendChild(searchInput);
+        searchInputWrapper.appendChild(clearButton);
+        
+        // Add close button to search container
         const closeButton = document.createElement('button');
         closeButton.className = 'model-selector-close';
         closeButton.innerHTML = '<i class="fas fa-times"></i>';
         
         // Add elements to container
-        searchContainer.appendChild(searchInput);
+        searchContainer.appendChild(searchInputWrapper);
         searchContainer.appendChild(closeButton);
         modelSelectorPopup.appendChild(searchContainer);
         
@@ -128,6 +143,13 @@ document.addEventListener('DOMContentLoaded', async () => {
         searchInput.addEventListener('input', function() {
             const searchTerm = this.value.toLowerCase();
             updateModelsList(searchTerm);
+        });
+        
+        // Add clear button event listener
+        clearButton.addEventListener('click', function() {
+            searchInput.value = '';
+            searchInput.focus();
+            updateModelsList('');
         });
         
         closeButton.addEventListener('click', function() {

